include/zephyr/device.h

```diff
@@ -961,15 +966,15 @@ static inline bool z_impl_device_is_ready(const struct device *dev)
  * @param level Initialization level.
  * @param prio Initialization priority.
  */
-#define Z_DEVICE_INIT_ENTRY_DEFINE(node_id, dev_id, init_fn_, level, prio)     \
-	Z_DEVICE_LEVEL_CHECK_DEPRECATED_LEVEL(level)                           \
-                                                                               \
-	static const Z_DECL_ALIGN(struct init_entry) __used __noasan           \
-		Z_INIT_ENTRY_SECTION(level, prio,                              \
-				     Z_DEVICE_INIT_SUB_PRIO(node_id))          \
-		Z_INIT_ENTRY_NAME(DEVICE_NAME_GET(dev_id)) = {                 \
-			.init_fn = {.dev = (init_fn_)},                        \
-			.dev = &DEVICE_NAME_GET(dev_id),                       \
+#define Z_DEVICE_INIT_ENTRY_DEFINE(node_id, dev_id, init_fn_, level, prio)                         \
+	Z_DEVICE_LEVEL_CHECK_DEPRECATED_LEVEL(level)                                               \
+                                                                                                   \
+	static const Z_DECL_ALIGN(struct init_entry) __used __noasan Z_INIT_ENTRY_SECTION(         \
+		level, prio, Z_DEVICE_INIT_SUB_PRIO(node_id))                                      \
+		Z_INIT_ENTRY_NAME(DEVICE_NAME_GET(dev_id)) = {                                     \
+			.init_fn = {COND_CODE_1(Z_DEVICE_IS_MUTABLE(node_id), (.dev_rw), (.dev)) = \
+					    (init_fn_)},                                           \
```

Comment:
This initialization syntax seems to break RG-2017.8-linux/XtensaTools/bin/xt-xcc  version 12.0.8, see below.

That toolchain is admittedly old (based on GCC 4 or 5) but it's a pity that an "experimental" feature breaks the build of every .c file even when none of them uses that new feature... is there any C99 compliant way to do the same?

```
13:03:03 zephyr/subsys/logging/log_core.c:944: error: unknown field ‘dev’ specified in initializer
13:03:03 zephyr/subsys/logging/log_core.c:944: warning: missing braces around initializer
13:03:03 zephyr/subsys/logging/log_core.c:944: warning: (near initialization for ‘__init_enable_logger.
```

cc: @dcpleung , @andyross , @kv2019i

Reply:
@marc-hb - can you submit a PR? It's unlikely I will have a similar toolchain at my disposal to test.

Reply:
Absolutely; I'd love to test and submit a change but first I need to understand which "advanced" (> C99) struct initialization feature is at play here. Either I'm a C99 dinosaur or the "macrobatics" confuse me or most likely: a bit of both. Care to share a feature name / search keyword / C version this initialization style relies on? If nothing else, an example of how the code looks after macro expansion. Any clue appreciated.

Reply:
`warning: missing braces around initializer` could it need an extra pair of braces around `init_fn_`? `COND_CODE_1` always needs extra ones, never understood why, wondering if if they get eaten up by the macro processor.

Reply:
the expansion for me is exactly the same before and after btw

`__attribute__((__section__( ".z_init_" "POST_KERNEL" "90""_" "00019""_"))) __init___device_dts_ord_19 = { .init_fn = {.dev = (&led_gpio_init)}, .dev = &__device_dts_ord_19, };` (this is `build/zephyr/drivers/led/CMakeFiles/drivers__led.dir/led_gpio.c.i` with `COMPILER_SAVE_TEMPS=y`)